### ISSUE TITLE
Setting terminal size of a sub process

### ISSUE THREAD

**Issue by xuoguoto** [ronf/asyncssh#657]
Hello all,

This is a follow up to an earlier question about asyncssh and cmd2 at #648 .

I have an ssh server with following code:
```python
import asyncio, asyncssh, subprocess, sys
import logging, os

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)

class MySFTPServer(asyncssh.SFTPServer):
    def __init__(self, chan: asyncssh.SSHServerChannel):
        root = '/tmp/'
        os.makedirs(root, exist_ok=True)
        super().__init__(chan, chroot=root)

async def handle_client(process: asyncssh.SSHServerProcess) -> None:
    bc_proc = subprocess.Popen('./script.sh', shell=True, stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    await process.redirect(stdin=bc_proc.stdin, stdout=bc_proc.stdout,
                           stderr=bc_proc.stderr)
    await process.stdout.drain()
    process.exit(0)

async def start_server() -> None:
    await asyncssh.listen('127.0.0.1', 8022, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',line_editor=False,
                          process_factory=handle_client, sftp_factory=MySFTPServer,
                          allow_scp=True)

loop = asyncio.get_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))

loop.run_forever()
```

This invokes a shell script which is as follows:

```bash
#!/bin/sh                                                                                                                               
/bin/script -q -c ./cmd01.py /tmp/transscript.log
exit 0
```
This invokes a simple cmd2 app

```python
#!/usr/bin/env python                                                                                                                   
"""A simple cmd2 application."""
import cmd2, os


class FirstApp(cmd2.Cmd):
    """A simple cmd2 application."""


if __name__ == '__main__':
    import sys
    print (os.get_terminal_size())
    c = FirstApp()
    sys.exit(c.cmdloop())
```

If you notice, in the python script I am printing the terminal size using `print (os.get_terminal_size())`. If I run `script.sh` in bash I get the following output:

```
$ ./script.sh 
os.terminal_size(columns=136, lines=43)
(Cmd) help

Documented commands (use 'help -v' for verbose/'help <topic>' for details):
===========================================================================
alias  help     macro  run_pyscript  set    shortcuts
edit   history  quit   run_script    shell

(Cmd)
```
Where I can see that the columns and lines are printed correctly. 

Now when I ssh into the asyncssh program, I get the following output:

```
$ ssh -p 8022  user@localhost 
os.terminal_size(columns=0, lines=0)
(Cmd) help

Documented commands (use 'help -v' for verbose/'help <topic>' for details):
===========================================================================
alias  help     macro  run_pyscript  set    shortcuts
edit   history  quit   run_script    shell

(Cmd) 
Connection to localhost closed.
```
The columns and lines are 0. 

If I see the debug logs of asyncssh:

```
DEBUG:asyncio:Using selector: EpollSelector
INFO:asyncssh:Creating SSH listener on 127.0.0.1, port 8022
INFO:asyncssh:[conn=0] Accepted SSH client connection
INFO:asyncssh:[conn=0]   Local address: 127.0.0.1, port 8022
INFO:asyncssh:[conn=0]   Peer address: 127.0.0.1, port 38922
<snip>
INFO:asyncssh:[conn=0] Auth for user user succeeded
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] New SSH session requested
DEBUG:asyncssh:[conn=0, chan=0]   Terminal type: screen.xterm-256color
DEBUG:asyncssh:[conn=0, chan=0]   Terminal size: 136x43
INFO:asyncssh:[conn=0, chan=0]   PTY created
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_ADDRESS=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_NAME=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_MONETARY=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LANG=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_PAPER=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_IDENTIFICATION=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_TELEPHONE=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_MEASUREMENT=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_TIME=en_IN
DEBUG:asyncssh:[conn=0, chan=0]   Env: LC_NUMERIC=en_IN
INFO:asyncssh:[conn=0, chan=0]   Interactive shell requested
INFO:asyncssh:[conn=0, chan=0] Sending exit status 0
INFO:asyncssh:[conn=0, chan=0] Closing channel
INFO:asyncssh:[conn=0, chan=0] Received channel close
DEBUG:asyncssh:[conn=0] Received disconnect: disconnected by user (11)
INFO:asyncssh:[conn=0, chan=0] Channel closed
```
I can see that the server is getting correct terminal size as 136x43.

But the client side terminal size not correctly set. I _think_ the pty is running in an 80x24 terminal, as, if I run vim in the cmd, the screen size is seen as 80x24. 

How can I set the terminal size of the script to correct terminal size?

**Comment by pyhedgehog** [comment]
Problem is here:
```
    bc_proc = subprocess.Popen('./script.sh', shell=True, stdin=subprocess.PIPE,
                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
You are starting your script with stdin/stdout redirected to newly created pipes (not a tty).

**Comment by ronf** [comment]
Redirecting the original stdin/stdout/stderr might be an option if you don't want your application to capture or modify any of the input or output. Also, if you do this, you'll probably only be able to get a single command to run out of this,  or it will require some special handling to safely do your own I/O on these streams when the upstream script exits, since it will try and close those streams.

You could also use the "pty" module in Python to open up your own pty to use for the upstream script, and then funnel data between the SSH session and that upstream PTY. I haven't ever tried this, so I'm not sure if there are any issues with that, but it would give you total control over the input & output.

**Comment by xuoguoto** [comment]
I have tried with `pty` module also:
```python
#!/usr/bin/env python                                                                                                                           
import os
import pty

shell = "./cmd01.py"

def read(fd):
    data = os.read(fd, 1024)
    return data

pty.spawn(shell, read)

```

This also works, and I am able to execute commands continuously. My only problem is that I have no means to set the size of the pty to match the size of window I am working on

**Comment by xuoguoto** [comment]
> Problem is here:
> 
> ```python
>     bc_proc = subprocess.Popen('./script.sh', shell=True, stdin=subprocess.PIPE,
>                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
> ```
> 
> You are starting your script with stdin/stdout redirected to newly created pipes (not a tty). You can get same effect in shell:

In this case I am able to work correctly (send commands and receive their reply multiple times) with the remote cmd program via ssh. As I said in previous message, my only problem is that the pty size is fixed and do not change with the window size of my ssh client.

I can see that the asyncssh does get correct window size from the client, but I have no means to pass that to underlying pty.

**Comment by ronf** [comment]
When you create your pty with pty.openpty(), you'll get back a pair of file descriptors. The first is the master side of the pty, which you would want to redirect the SSH connection to read & write to and the second is the slave side of the pty which you would want to redirect the Popen to redirect to.

To set the terminal size, you'd need to do an ioctl of TIOCSWINSZ on the master side of the pty before you began doing the redirection, passing in the terminal size you get from AsyncSSH. The code would look something like:

```python
    fcntl.ioctl(pty_master, termios.TIOCSWINSZ, struct.pack('hhhh', width, height, pixwidth, pix height))
```

If you need to handle window size changes, AsyncSSH provides a way to get an exception when the terminal is resized, and you could redo your ioctl() to pass that through to the pty when that happens.

**Comment by ronf** [comment]
Here's a more complete example:

```python
async def handle_connection(process):
    local_pty, local_tty = os.openpty()

    fcntl.ioctl(local_pty, termios.TIOCSWINSZ,
                struct.pack('hhhh', *process.term_size))

    local_proc = subprocess.Popen('stty -a', shell=True,
                                  stdin=local_tty, stdout=local_tty)
    os.close(local_tty)

    await process.redirect(stdin=local_pty, stdout=local_pty)
    await process.stdout.drain()
    process.exit(0)
```

In this version I used `os.openpty()` instead of `pty.openpty()`, since they're basically equivalent and this can save the need for an extra import.

Note that I didn't do anything with stderr here on either the remote subprocess or on the SSH channel redirect. This is because PTYs only ever communicate through stdin & stdout. There is no separate stderr stream when a PTY is allocated.

Also, note there's no explicit close of local_pty. This is handled automatically by the SSH process redirection. For local_tty, though, the subprocess will close the copy of the fd in the subprocess, but it won't close it in the parent, so that has to be done explicitly. This can be done right after the Popen creates the subprocess, as the copy of the fd in the subprocess will exist at that point and will stay open until the subprocess exits. Closing the parent fd allows us to get an EOF notification when the subprocess exits (waking up the call to `drain`).

**Comment by xuoguoto** [comment]
Thanks for the detailed reply, made some changes and the initial terminal size is drawn as expected.

In the example height and width were interchanged, that was corrected.
```py
fcntl.ioctl(pty_master, termios.TIOCSWINSZ, struct.pack('hhhh', width, height, pixwidth, pix height))
```
for ioctl `height` should be first and `width` second, like
```py
width, height, pixWidth, pixHeight = process.term_size
fcntl.ioctl(local_tty, termios.TIOCSWINSZ, struct.pack('hhhh', height, width, pixHeight, pixWidth))
```

Also the `pty` and `tty` returned from `os.openpty()` were interchanged, for subprocess and ioctl `tty` was used instead of `pty` 
for redirect `pty` was used instead of `tty`. 

here is the complete program:
```py
import asyncio, asyncssh, sys, subprocess, os, fcntl, struct, termios, logging

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)

async def handle_client(process: asyncssh.SSHServerProcess) -> None:

    local_pty, local_tty = os.openpty()
    width, height, pixWidth, pixHeight = process.term_size
    fcntl.ioctl(local_tty, termios.TIOCSWINSZ,
                struct.pack('hhhh', height, width, pixHeight, pixWidth))

    local_proc = subprocess.Popen('./script.sh', shell=True,
                                  stdin=local_pty, stdout=local_pty)
    os.close(local_pty)

    await process.redirect(stdin=local_tty, stdout=os.dup(local_tty))
    await process.stdout.drain()
    process.exit(0)


async def start_server() -> None:
    await asyncssh.listen('127.0.0.1', 8022, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',line_editor=False,
                          process_factory=handle_client)

loop = asyncio.get_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))
try:
    loop.run_forever()
except KeyboardInterrupt as exc:
    sys.exit('Program stopped')```

But  resizing doesn't work though.

**Comment by xuoguoto** [comment]
>  If you put back the associations with just the width & height swapped, does that work for you?

Without swapping PTY and TTY the client wasn't disconnecting properly. The only way to disconnect the client was to stop the server without the swap, that's why they were swapped.

> I'll experiment with this and get back to you.

Thanks! Will wait for your update.

**Comment by ronf** [comment]
Ok - I've got a prototype for forwarding window size changes and break from an SSH connection to a local TTY:

```patch
diff --git a/asyncssh/process.py b/asyncssh/process.py
index 7c7b5a9..a6b46b5 100644
--- a/asyncssh/process.py
+++ b/asyncssh/process.py
@@ -29,6 +29,7 @@ import os
 from pathlib import PurePath
 import socket
 import stat
+import sys
 from types import TracebackType
 from typing import Any, AnyStr, Awaitable, Callable, Dict, Generic, IO
 from typing import Iterable, List, Mapping, Optional, Set, TextIO
@@ -51,6 +52,11 @@ from .stream import SSHReader, SSHWriter, SSHStreamSession
 from .stream import SSHClientStreamSession, SSHServerStreamSession
 from .stream import SFTPServerFactory

+if sys.platform != 'win32': # pragma: no branch
+    import fcntl
+    import struct
+    import termios
+

 _AnyStrContra = TypeVar('_AnyStrContra', bytes, str, contravariant=True)

@@ -434,6 +440,17 @@ class _PipeWriter(_UnicodeWriter[AnyStr], asyncio.BaseProtocol):
         assert self._transport is not None
         self._transport.write(self.encode(data))

+    def write_exception(self, exc: Exception) -> None:
+        fd = self._transport.get_extra_info('pipe').fileno()
+
+        if sys.platform != 'win32' and os.isatty(fd):
+            if isinstance(exc, TerminalSizeChanged):
+                fcntl.ioctl(fd, termios.TIOCSWINSZ,
+                            struct.pack('hhhh', exc.height, exc.width,
+                                        exc.pixwidth, exc.pixheight))
+            elif isinstance(exc, BreakReceived): # pragma: no branch
+                termios.tcsendbreak(fd, exc.msec)
+
     def write_eof(self) -> None:
         """Write EOF to the pipe"""

```

Passing through signals would be trickier, as that can't be done though a pipe. It would only be possible if a local subprocess was provided as the target of a redirect so that AsyncSSH had access to the pid. So, for now, I've left that out.

Passing through break seems to work, but it doesn't appear to actually do anything when sent to a shell, even when BRKINT is set (which should cause a break to turn into an interrupt, like ^C). Given that, I may leave this out and only handle the window size change. Let me know if it works for you.

One other thing I noticed that is that you probably should be redirecting both stdout and stderr in the local subprocess call to the TTY. Otherwise, stdout will continue to go to the stdout of the AsyncSSH server process. I ended up doing the following for this:

```python
    local_proc = subprocess.Popen('stty -a', shell=True, stdin=local_tty,
                                  stdout=local_tty, stderr=local_tty)
```

Note that this only applies to the local_proc and not to the stdout/stderr on the AsyncSSH side (since making that a PTY merges together stdout & stderr). Also, it's important to make sure the os.dup() call is there. Otherwise, you can get errors when cleaning things up.

Once I added the call to os.close(local_tty), things started to exit properly for me.

**Comment by ronf** [comment]
A somewhat more comprehensive version of the above is now available in the "develop" branch as commit 1bbd845. This version will automatically set the initial terminal size provided by the client when the SSHServerProcess redirects stdin to a local TTY, in addition to passing through any size updates after that.

Here's an updated example client. Note that there's no direct terminal size code here any more, as AsyncSSH is now taking care of that.

```python
import asyncio
import asyncssh
import os
import subprocess
import sys

async def handle_connection(process):
    local_pty, local_tty = os.openpty()

    local_proc = subprocess.Popen('while true; do stty size; sleep 5; done',
                                  shell=True, stdin=local_tty,
                                  stdout=local_tty, stderr=local_tty)
    os.close(local_tty)

    await process.redirect(stdin=local_pty, stdout=os.dup(local_pty))

    await process.stdout.drain()
    process.exit(0)

async def start_server():
    await asyncssh.listen('', 8022, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',
                          process_factory=handle_connection)

loop = asyncio.new_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))

loop.run_forever()
```

**Comment by ronf** [comment]
Actually, you may want to hold off a bit on trying this. I'm seeing some unit test failures on Linux and Windows that I didn't see on macOS. I'll let you know when this is resolved.

**Comment by ronf** [comment]
Ok - commit 43bcd2d should address this.

**Comment by xuoguoto** [comment]
Thanks, I was checking with the old code and was facing some difficulty. Will check this out and let you know.

**Comment by ronf** [comment]
Regarding the Event.wait message, that can happen if you exit the Python interpreter without giving the various cleanup functions a chance to run. In this case, I think the problem might be that the local proc continues to run even after the client closes the SSH connection. So, there's still an instanced of `handle_connection()` running that gets destroyed without cleanup when the server is stopped. If the server-side process exited when it saw EOF on stdin, I don't think you'd see this problem.

Also, it looks like my example code is missing a call to `await process.wait_closed()` at the end of `handle_connection()`, or putting the entire body of that function in an `async with process:` block. That may be needed as well to avoid an event object there from not being properly awaited when `handle_connection()` exits.

As for the rest of what you've described, most of it sounds out of AsyncSSH's control. All it can do is to do the appropriate ioctl() to tell the pseudo-TTY that the size has changed. Everything else is up to the application running on that TTY.

Regarding switching pty & tty, do you have a minimal example where you can reproduce this problem? If I change the example above to run a command which exits, the SSH connection from the client is closed when the command finishes.

**Comment by xuoguoto** [comment]
> > Also I'm still having trouble without switching pty and tty as the client can't be disconnected even when the subprocess is complete, only way to disconnect the client is to stop the server. Switching tty and pty will disconnect the client when the subprocess is complete/exited.

> Regarding switching pty & tty, do you have a minimal example where you can reproduce this problem? If I change the example above to run a command which exits, the SSH connection from the client is closed when the command finishes.

I am attaching a minimal example where this problem is illustrated. Also a short video of the experience when running this.
Screen cast of running these two programs: [Screencast.webm](https://github.com/ronf/asyncssh/assets/79757767/4eb5423a-33b9-49a4-ad0e-ba4241c3c1ab)
 
I am pasting the programs here as gh do not allow me to attach these files.

`cmd01.py`
```py
#!/usr/bin/env python
"""A simple cmd2 application."""
import cmd2, os

class FirstApp(cmd2.Cmd):
    """A simple cmd2 application."""

if __name__ == '__main__':
    import sys
    c = FirstApp()
    sys.exit(c.cmdloop())
```

`test.py`
```py
import asyncio, sys, subprocess, os, fcntl, struct, termios, logging, asyncssh

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)

async def handle_client(process: asyncssh.SSHServerProcess) -> None:

    local_pty, local_tty = os.openpty()
    # width, height, pixWidth, pixHeight = process.term_size
    # fcntl.ioctl(local_pty, termios.TIOCSWINSZ,
    #             struct.pack('hhhh', height, width, pixWidth, pixHeight))

    local_proc = subprocess.Popen('./cmd01.py', shell=True,
                                  stdin=local_tty, stdout=local_tty, stderr=local_tty)
    os.close(local_tty)

    await process.redirect(stdin=local_pty, stdout=os.dup(local_pty))

    await process.stdout.drain()
    await process.wait_closed()
    process.exit(0)


async def start_server() -> None:
    await asyncssh.listen('127.0.0.1', 8022, server_host_keys=['ssh_host_key'],
                          authorized_client_keys='ssh_user_ca',line_editor=False,
                          process_factory=handle_client)

loop = asyncio.get_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))
try:
    loop.run_forever()
except KeyboardInterrupt as exc:
    sys.exit('Program stopped')
```

**Comment by xuoguoto** [comment]
I had made this change, but there seems to be one more issue. The `await process.stdout.drain()` is never completed so call to `process.exit(0)` is never reached.  I had put a log line immediately after `await process.stdout.drain()` and that line is never printed. I am not sure if this is some platform specefic issue as it seems to be running fine at your end.

**Comment by ronf** [comment]
Is stdin still not returning from drain() even when you type "exit" into the Cmd shell? Once all current output is read and EOF is received on a particular data channel, the redirection should be disabled and then `drain()` should return. What does your debug output look like after entering "exit"?

**Comment by xuoguoto** [comment]
Just to recap, I am testing two variations of the same scenario, one called `test-with-pty.py` and other `test-with-tty.py`. Difference between them is as follows:

```diff
--- test-with-pty.py    2024-05-31 11:38:43.351902272 +0530
+++ test-with-tty.py    2024-05-31 11:38:44.907893967 +0530
@@ -8,11 +8,11 @@
 
     local_pty, local_tty = os.openpty()
     
-    local_proc = subprocess.Popen('./cmd01.py', shell=True,
-                                  stdin=local_tty, stdout=local_tty, stderr=local_tty)
-    os.close(local_tty)
+    local_proc = subprocess.Popen('./script.sh', shell=True,
+                                  stdin=local_pty, stdout=local_pty, stderr=local_pty)
+    os.close(local_pty)
 
-    await process.redirect(stdin=local_pty, stdout=os.dup(local_pty))
+    await process.redirect(stdin=local_tty, stdout=os.dup(local_tty))
     await process.stdout.drain()
     process.exit(0)
     await process.wait_closed()
```

`test-with-pty.py` is is as per documentation with `redirect(stdin=local_pty, stdout=os.dup(local_pty))` while `test-with-tty.py` is pty and tty switched, ie `redirect(stdin=local_tty, stdout=os.dup(local_tty))`. 

Case 1. In `test-with-pty` I can directly open the `cmd01.py` resizing is working as expected, but the exit is not registered. I have attached the screen capture of how this goes.

[Case 1 Screen cast](https://github.com/ronf/asyncssh/assets/79757767/ff79d4d0-e740-4afb-8ab0-1eb40359b845)

Case 2. In `test-with-tty` if I directly open the `cmd01.py` (or another program like `bc`) the client program (`cmd2` in this case) goes into a loop. I have attached the screen capture. 

[Case 2 Screen cast](https://github.com/ronf/asyncssh/assets/79757767/93f1d5e1-5fa9-405a-bdfa-2a19e91e3cd4)

Case 3. In `test-with-tty` if use another script to run `cmd01.py` things are much better, It is disconnecting correctly and the only issue is that when height is changed, its getting updated only when width is changed. I have attached the video of this also.

[Case 3 Screen cast](https://github.com/ronf/asyncssh/assets/79757767/815b5d31-90c9-4f25-b336-fef0036d9e8c)

So this is the current status. I am testing this on python version :  3.10.12 on Ubuntu 22.04. 

> What does your debug output look like after entering "exit"?

This scenario happens in Case 1. After `exit` is pressed, there are no logs in the sever side. It just waits as you can see in the screen cast. Is there any additional debugging that can be done (for eg `tcpdump`) that will help? Also please note that it works as expected in Case 2.


[script.sh.txt](https://github.com/user-attachments/files/15509830/script.sh.txt)
[test-with-tty.py.txt](https://github.com/user-attachments/files/15509832/test-with-tty.py.txt)
[test-with-pty.py.txt](https://github.com/user-attachments/files/15509833/test-with-pty.py.txt)

**Comment by ronf** [comment]
Thanks for the additional info. I re-ran my test here on Ubuntu (previous runs were on macOS), and I was able to reproduce the hang in my original code (similar to your case 1). As you said, it seems to hang on `await process.stdout.drain()`, even after the Cmd shell exits. In fact, it remains hung there even after the SSH client disconnects, though there is debug output in that case showing that the server did process the connection close. I haven't yet had a chance to dig into why drain is hanging, but the OS difference appears to explain why you were seeing it and I wasn't. I'll take a closer look at this tonight.

On a separate note, your "case 3" screen capture doesn't appear to be doing anything. Was that the capture you intended to post?

**Comment by ronf** [comment]
Thanks for the confirmation. A slightly updated version of this fix is now available as commit 9ad9688 in the "develop" branch. It corrects an issue I discovered after running the unit tests.

**Comment by ronf** [comment]
I did some more investigation tonight.

On macOS, closing the slave TTY causes the master (PTY) to receive an EOF. However, it appears that on Linux, the master gets back a connection_lost with an EIO error, without sending EOF first, and the connection_lost code path  doesn't properly wake up readers that were blocked waiting for the redirect to complete.

I'll need some time to go through the code and find the best place to unblock waiters blocked on drain when a redirected pipe is being cleaned up. This difference in error handling would explain what we've been seeing, though, and I don't think it'll be too hard to fix.

**Comment by xuoguoto** [comment]
Thank you for your time and effort. I will wait for your investigation to be complete.

**Comment by ronf** [comment]
Ok - I have a proposed fix you can try:

```patch
diff --git a/asyncssh/process.py b/asyncssh/process.py
index bc3496d..3510e33 100644
--- a/asyncssh/process.py
+++ b/asyncssh/process.py
@@ -365,6 +365,12 @@ class _PipeReader(_UnicodeReader[AnyStr], asyncio.BaseProtocol):

         self._transport = cast(asyncio.ReadTransport, transport)

+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        """Handle closing of the pipe"""
+
+        self._process.feed_close(self._datatype)
+        self.close()
+
     def data_received(self, data: bytes) -> None:
         """Forward data from the pipe"""

@@ -1048,6 +1054,12 @@ class SSHProcess(SSHStreamSession, Generic[AnyStr]):
         self._readers[datatype].close()
         self.clear_reader(datatype)

+    def feed_close(self, datatype: DataType) -> None:
+        """Feed pipe close to the channel"""
+
+        if datatype in self._readers:
+            self.feed_eof(datatype)
+
     def feed_recv_buf(self, datatype: DataType,
                       writer: _WriterProtocol[AnyStr]) -> None:
         """Feed current receive buffer to a newly set writer"""
```

This change looks for cases where `connection_lost()` is called on a PipeReader without calling `eof_received()` first, and triggers the code which would have happened in `eof_received()` in that case. Among other things, this will disable the redirection and wake up any tasks blocked on `drain()`.

This seems to solve the problem on Ubuntu, and also continues to work on macOS (where `eof_received()` is called).

**Comment by xuoguoto** [comment]
Thanks, tried this out and it seems to fix the disconnection issue without swapping pty and tty. Thanks a lot for your patience and help!

**Comment by ronf** [comment]
Thanks for confirming, and for all your help tracking this down! This change is now in the "develop" branch as commit 5159542.

**Comment by imgurbot12** [comment]
Sorry to necro this thread, but I ran into major issues getting resizing to work properly when running a tui application in the subprocess. Turns out, the `TIOCSWINSZ` was being sent properly, but the application was looking for a `signal.SIGWINCH` to know when to make a redraw of the resize. I had to monkey-patch the exception handler to get it resize properly:

```python
local_pty, local_tty = os.openpty()

proc = subprocess.Popen(args=command,
    stdin=local_tty, stdout=local_tty, stderr=local_tty)
os.close(local_tty)

# monkey-patch exception handler to include SIGWINCH on resize
func = process.exception_received
def exception_received(exc: Exception):
    func(exc)
    if isinstance(exc, asyncssh.TerminalSizeChanged):
        proc.send_signal(signal.SIGWINCH)
process.exception_received = exception_received

await process.redirect(stdin=local_pty, stdout=os.dup(local_pty))
await process.stdout.drain()
process.exit(0)
```

Hopefully that helps somebody else avoid my immense pain in discovering this. Thanks! :)

**Comment by ronf** [comment]
The main challenge I see here with triggering the SIGWINCH on the server side is that AsyncSSH only has stdin/stdout connected to a PTY here. That's enough to do the fcntl.ioctl() to communicate the terminal size, but not enough to send a signal. For that, you need access to the subprocess instance.

Rather than doing redirection, you could potentially keep your own read/write loops which would catch TerminalSizeChanged and do both the set_terminal_size() as well as sending SIGWINCH to the server-side subprocess. It'd be nice if there was a way to let redirect handle all the data and exception forwarding, though. Perhaps an optional pid could be passed when setting up redirects, similar to what is being done in the PipeWriter for remembering the tty (through get_extra_info).

Out of curiosity, if you do get_extra_info('subprocess') on the transport given to _PipeWriter's connection_made(), does it have a value? I doubt it, since we're just passing in TTYs right now, but it's worth a quick look.

Do you have a runnable test case I could use to experiment with this?

**Comment by imgurbot12** [comment]
@ronf 

> Out of curiosity, if you do get_extra_info('subprocess') on the transport given to _PipeWriter's connection_made(), does it have a value? I doubt it, since we're just passing in TTYs right now, but it's worth a quick look.

Unfortunately not. I ran a local copy of asyncssh and modified the source as you mentioned to try and it just comes back with `None`.

<details>
  <summary>Here are the results</summary>

```python
def connection_made(self, transport: asyncio.BaseTransport) -> None:
    """Handle a newly opened pipe"""
  
    self._transport = cast(asyncio.WriteTransport, transport)
  
    pipe = transport.get_extra_info('pipe')
    print('SUBPROCESS!', transport.get_extra_info('subprocess'))
  
    if isinstance(self._process, SSHServerProcess) and pipe.isatty():
        self._tty = pipe
        set_terminal_size(pipe, *self._process.term_size)
  ```

  `SUBPROCESS! None`
</details>

> Do you have a runnable test case I could use to experiment with this?

Sure! I put together a POC for testing. I'm running a [textual](https://textual.textualize.io/) app as the subprocess service being passed over ssh:

<details>
  <summary>app.py</summary>

```python
from textual.app import App, ComposeResult
from textual.widgets import Static

class HelloWorld(App):

    def compose(self) -> ComposeResult:
        yield Static("""
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
        Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
        Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.""" * 3)

HelloWorld().run()
```
</details>

<details>
  <summary>ssh.py</summary>

```python
import os
import sys
import asyncio
import logging
import signal
import subprocess

import asyncssh

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)

class Server(asyncssh.SSHServer):
    def begin_auth(self, username: str) -> bool:
        return False

async def handle_client(process: asyncssh.SSHServerProcess) -> None:
    local_pty, local_tty = os.openpty()

    local_proc = subprocess.Popen('python3 ./app.py', shell=True,
        stdin=local_pty, stdout=local_pty, stderr=local_pty)
    os.close(local_pty)

    # monkey-patch exception handler to include SIGWINCH on resize
    func = process.exception_received
    def exception_received(exc: Exception):
        func(exc)
        if isinstance(exc, asyncssh.TerminalSizeChanged):
            local_proc.send_signal(signal.SIGWINCH)
    process.exception_received = exception_received

    await process.redirect(stdin=local_tty, stdout=os.dup(local_tty))
    await process.stdout.drain()
    process.exit(0)
    await process.wait_closed()

async def start_server() -> None:
    key = asyncssh.generate_private_key('ssh-rsa')
    key.write_private_key('ssh_host_key')

    server = await asyncssh.create_server(Server,
        '127.0.0.1',
        8022,
        server_host_keys=['ssh_host_key'],
        line_editor=False,
        process_factory=handle_client)
    try:
        await server.wait_closed()
    finally:
        server.close()

asyncio.run(start_server())
```
</details>

Run the example by calling `python ssh.py` with `app.py` in the same working directory.
Hope it helps. Thanks for the response! :)

**Comment by ronf** [comment]
Thanks very much for the test code! I've been able to reproduce the original issue with the delayed screen redraw, and see how your patch of exception_received() fixes this.

My current thinking is to add something like `set_redirect_pid()` as a method on `SSHServerProcess`. This could be set in `handle_client()` after starting the subprocess. Later, where `set_terminal_size()` is called from within `_PipeWriter`, it could check if the pid is set and trigger the SIGWINCH an after resetting the terminal size. The process object is already accessible from there, so getting the pid shouldn't be a problem (if one was set).

It's bit inelegant to set the pid manually like this, but the only other way I could see to do this would be to add a special version of a subprocess object for this, but right now that's only supported on the client side, to start up remote subprocesses on the SSH server and tunnel data back. There is no equivalent specialized class for handling incoming subprocess requests to a server, as they just look like regular shell/exec/subsystem calls there.

On an unrelated note, have you noticed that quitting out the textual app seems to leave the terminal in a broken state related to event delivery, even when exiting "cleanly" with Ctrl-Q?

**Comment by imgurbot12** [comment]
Glad to hear it! Anytime! Really appreciate the follow up :)

That sounds like a fine solution. I wouldn't claim to know what would probably work best for the library since this is my first attempt at a project utilizing it. The "custom" subprocess caller might not be terrible either if you're willing to do that. 

As someone who doesn't know the inns and outs of terminal protocols and handling very well, the opening of a pty/tty with that os command and then redirecting it is all mostly magic to me thus far, so something that just abstracts it all into a single `forward_subprocess` call with all the logic baked in or something could be pretty nice for laymen like myself.

I cannot say I have encountered any issues with the terminal being broken thus far, whether exiting cleanly or otherwise. I'm using wezterm which is a bit wonky so maybe I'm just lucky lol.

**Comment by ronf** [comment]
Ok - I have something you can try:

```patch
diff --git a/asyncssh/process.py b/asyncssh/process.py
index bf58ca5..889a65b 100644
--- a/asyncssh/process.py
+++ b/asyncssh/process.py
@@ -27,6 +27,7 @@ import inspect
 import io
 import os
 from pathlib import PurePath
+import signal
 import socket
 import stat
 from types import TracebackType
@@ -461,11 +462,20 @@ class _PipeWriter(_UnicodeWriter[AnyStr], asyncio.BaseProtocol):
         self._transport.write(self.encode(data))

     def write_exception(self, exc: Exception) -> None:
-        """Write terminal size changes to the pipe if it is a TTY"""
-
-        if isinstance(exc, TerminalSizeChanged) and self._tty:
+        """Write signals and terminal size changes to the pipe"""
+
+        if isinstance(exc, BreakReceived):
+            if self._process.redirect_pid:
+                os.kill(self._process.redirect_pid, signal.SIGINT)
+        elif isinstance(exc, SignalReceived):
+            if self._process.redirect_pid:
+                os.kill(self._process.redirect_pid, exc.signal)
+        elif isinstance(exc, TerminalSizeChanged) and self._tty:
             set_terminal_size(self._tty, *exc.term_size)

+            if self._process.redirect_pid:
+                os.kill(self._process.redirect_pid, signal.SIGWINCH)
+
     def write_eof(self) -> None:
         """Write EOF to the pipe"""

@@ -1593,6 +1603,7 @@ class SSHServerProcess(SSHProcess[AnyStr], SSHServerStreamSession[AnyStr]):

         self._process_factory = process_factory

+        self._redirect_pid: Optional[int] = None
         self._stdin: Optional[SSHReader[AnyStr]] = None
         self._stdout: Optional[SSHWriter[AnyStr]] = None
         self._stderr: Optional[SSHWriter[AnyStr]] = None
@@ -1644,6 +1655,12 @@ class SSHServerProcess(SSHProcess[AnyStr], SSHServerStreamSession[AnyStr]):

         return self._chan.get_terminal_modes()

+    @property
+    def redirect_pid(self) -> Optional[int]:
+        """The pid that signals and terminal changes should be sent to"""
+
+        return self._redirect_pid
+
     @property
     def stdin(self) -> SSHReader[AnyStr]:
         """The :class:`SSHReader` to use to read from stdin of the process"""
@@ -1777,6 +1794,11 @@ class SSHServerProcess(SSHProcess[AnyStr], SSHServerStreamSession[AnyStr]):

         await self.redirect(None, None, source, bufsize, send_eof, True)

+    def set_redirect_pid(self, pid: Optional[int]) -> None:
+        """Set the pid that signals and terminal changes should be sent to"""
+
+        self._redirect_pid = pid
+
     def get_terminal_type(self) -> Optional[str]:
         """Return the terminal type set by the client for the process

```

The idea of having a wrapper which automatically sets up a PTY and redirects to a subprocess is kind of interesting, and that could also handle setting the new redirect PID as part of that. I'll dig into that a bit more tomorrow.

Regarding the terminal getting into a weird state, I was using iTerm 2. I also see a version of the problem with the macOS Terminal.app, but it's not quite as severe. At least you can still enter input in the Terminal.app case. With iTerm your only option is to restart the terminal window, as even regular input characters appear to be turned into escape sequences. I think this might have to do with wanting to distinguish key down and key up events in textual, as well as other things like mouse position and button presses, and focus events. I really would have expected textual to reset all those setting on a clean exit, though.

**Comment by imgurbot12** [comment]
> Ok - I have something you can try:

Oh noice. It works well! SIGWINCH passed and it resizes properly. Thanks :)

> The idea of having a wrapper which automatically sets up a PTY and redirects to a subprocess is kind of interesting, and that could also handle setting the new redirect PID as part of that. I'll dig into that a bit more tomorrow.

I would be interested to see what you come up with if you decide to go that direction. It would be slightly more "ergonomic" for my use case since I'm basically just forwarding the client over ssh, so a single call would look and operate much more cleanly but maybe that's just me :)

> I think this might have to do with wanting to distinguish key down and key up events in textual, as well as other things like mouse position and button presses, and focus events. I really would have expected textual to reset all those setting on a clean exit, though.

Very interesting. I can't claim to know much about textual either since this is my first time playing around with it, but I have yet to encounter any issues personally. I tested the POC with wezterm, alacritty, rio-term, foot, and even xterm and did not seem to encounter that. Weird. Maybe its a MacOS thing? :/

**Comment by ronf** [comment]
> Very interesting. I can't claim to know much about textual either since this is my first time playing around with it, but I have yet to encounter any issues personally. I tested the POC with wezterm, alacritty, rio-term, foot, and even xterm and did not seem to encounter that. Weird. Maybe its a MacOS thing? :/

Even xterm fails for me on the Mac, though in a less severe way. The block cursor ends up disappearing after you exit the app, but you can still type normally.

The real question for me is whether the app is failing to send the escape sequences to turn off the special input mode, or whether something about the way we clean up the subprocess maybe doesn’t a chance to forward the escape sequences to the client. It’s possible that could be Mac-specific, somehow tied to the way the OS handles pseudo-TTYs, for instance.

If you run in debug mode level 3, I think you can actually see all the bytes going in and out of the SSH tunnel, and if we compare what that looks like between OSes, we might be able to figure out if this is an issue. Just be careful not to show any passwords that might appear in the trace if you play with this.

On the wrapper, I'm still thinking about whether it is worth it or not, and exactly what parameters would make sense to pass in to still provide a good amount of flexibility in what gets started up and when things like TTY and terminal information get passed through.

I'm also worried about some of the potential security concerns. For instance, I don't think we'd want to preserve the command requested by the client, if there was one.

**Comment by imgurbot12** [comment]
> If you run in debug mode level 3, I think you can actually see all the bytes going in and out of the SSH tunnel, and if we compare what that looks like between OSes, we might be able to figure out if this is an issue. Just be careful not to show any passwords that might appear in the trace if you play with this.

I published my log connecting to the POC of the following command to gist if its of any value to you: 
`ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null guest@localhost -p 8022 -vvv -E client.log`
https://gist.github.com/imgurbot12/31eb267b099e40d11339bda7e80e2725

> I'm also worried about some of the potential security concerns. For instance, I don't think we'd want to preserve the command requested by the client, if there was one.

Totally understand. I'm not sure I follow completely since like I said: I don't really know much about ssh beyond how to use it, but if its more hassle than its worth no worries.

**Comment by ronf** [comment]
Sorry, I meant AsyncSSH debug level 3:

```python
import logging
logging.basicConfig(level='DEBUG')
asuncssh.set_debug_level(3)
```

I don't think OpenSSH is capable of dumping raw data without recompiling it with debugging enabled.

If you do capture the AsyncSSH debug output, I would avoid pasting anything before "INFO:asyncssh:[conn=5] Auth for user <username> succeeded", so you don't accidentally include cleartext passwords in the output. In this case, the handshake is not relevant, so only including log entries after auth has succeeded should be fine.

**Comment by imgurbot12** [comment]
Oh my bad lol. I assumed you meant the client since it had the `-vvv`. I didn't realize asyncssh had such an option.

Here is a truncated server log with the same client command being used.
https://gist.github.com/imgurbot12/e8b5a8d2e900db82c9a8831ef74041df

**Comment by ronf** [comment]
Ok - cleanup escape sequences are definitely being sent right after you hit Ctrl-Q and before EOF is sent. The relevant data packet is:

```
DEBUG:asyncssh:[conn=0, chan=0, pktid=59] Sent MSG_CHANNEL_DATA (94), 54 bytes
  00000000: 5e 00 00 00 00 00 00 00 2d 1b 5b 3f 32 30 30 34  ^.......-.[?2004
  00000010: 6c 1b 5b 3f 37 68 1b 5b 3f 31 30 30 30 6c 1b 5b  l.[?7h.[?1000l.[
  00000020: 3f 31 30 30 33 6c 1b 5b 3f 31 30 31 35 6c 1b 5b  ?1003l.[?1015l.[
  00000030: 3f 31 30 30 36 6c                                ?1006l
```

This translates to:

```
ESC [ ? 2004 l    Disable bracketed paste mode
ESC [ ? 7 h       Re-enable auto wrap
ESC [ ? 1000 l    Disable mouse click and movement tracking
ESC [ ? 1003 l    Disable any-event mouse tracking
ESC [ ? 1015 l    Disable UTF-8 mouse tracking (for larger screens)
ESC [ ? 1006 l    Disable SGR extended mouse mode
```

It's basically a bunch of different variants of sequence to enable mouse tracking, since that can vary from one terminal emulator to another. Automatic line wrap is also re-enabled before textual shuts down.

Unfortunately, I can't tell from the SSH client log whether the bytes made it all the way through the tunnel and out to the client terminal, but I know for sure AsyncSSH is sending them.

**Comment by ronf** [comment]
The new `set_ridirect_pid()` method is now available in commit df0a544 in the "develop" branch and will be included in the next release.

I've got some other queued up work items here, so I'm going to put the wrapper we discussed on hold for now. I still need more time to think about if there's a good way to offer that while still keeping it fairly configurable. Even without the wrapper, it's only about 10 lines of code with `set_redirect_pid()`, so it probably wouldn't get too much shorter.

**Comment by imgurbot12** [comment]
> The new set_ridirect_pid() method is now available

Amazing! Thanks for all your support :)
